New JWT Vending Machine API (#891)
FusionAuth/fusionauth-site: site/docs/v1/tech/apis/jwt.adoc

=== Request

> // [brettp]Note: This says "Issue an Access Token" above it say "Issue a JWT", should we be consistent?

Reviewer: Yes, I vote for consistency.

[.api-authentication]

> link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Issue an Access Token

Reviewer: ```suggestion
link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Issue a JWT
```

== Retrieve Public Keys

> This API is used to retrieve Public Keys generated by FusionAuth, used to cryptographically verify JWT signatures signed using the corresponding RSA or ECDSA private key.

Reviewer: ```suggestion
This API is used to retrieve Public Keys generated by FusionAuth. These can be used to cryptographically verify JWTs signed with the corresponding RSA or ECDSA private key.
```

----
include::docs/src/json/jwt/validate-response.json[]
----

== Vend a JWT

> This API is used to create a free-form access token (JWT) with claims defined by the caller. The only "reserved" claims that cannot be specified are `exp`, `iat`. The `iat` the issued at time of the JWT and the `exp` the expiration time of the JWT as computed by adding either the user passed, or the tenant [field]#timeToLiveInSeconds# to the `iat`. If passed into the [field]#claims# object these claims will be ignored.

Reviewer: ```suggestion
This API is used to create a free-form access token (JWT) with claims defined by the caller. The only "reserved" claims that cannot be specified are `exp` and `iat`. The `iat` claim is the issued at time of the JWT and the `exp` is the expiration time of the JWT as computed by adding to the `iat` value either the user passed [field]#timeToLiveInSeconds# or the tenant JWT [field]#timeToLiveInSeconds#. If passed into the [field]#claims# object, these claims will be ignored.
```

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Vend a JWT
[.endpoint]
.URI
--
[method]#POST# [uri]#/api/jwt/vend#
--

include::docs/v1/tech/apis/_x-fusionauth-tenant-id-header-scoped-operation.adoc[]

==== Request Body

[.api]
[field]#claims# [type]#[Object]# [optional]#Optional#::

> An object of claims to apply to (associate with?) this JWT.  If any of the three "reserved" claims `exp`, `iat`, or `jti` are specified they will be ignored.

Reviewer: ```suggestion
An set of claims to add to this JWT.  If any of the "reserved" claims, `exp` or `iat`, are specified they will be ignored. Otherwise, the claims can be any valid JSON value.
```

[field]#keyId# [type]#[UUID]# [optional]#Optional#::

> The Id of the signing key to use when signing this JWT. If this is not supplied the signing key on the tenant will be used.

Reviewer: ```suggestion
The Id of the signing key to use when signing this JWT. If this is not supplied, the tenant's JWT access token signing key will be used.
```

[field]#timeToLiveInSeconds# [type]#[Integer]# [optional]#Optional#::

> The length of time in seconds the JWT will live before it is expired and no longer valid. Value must be greater than `0`. If omitted, the [field]#timeToLiveInSeconds# value on the tenant will be used instead.

Reviewer: ```suggestion
The length of time in seconds before the JWT expires and no longer valid. Any integer value greater than `0` is allowed. If omitted, the tenant's [field]#timeToLiveInSeconds# value will be used instead.
```
